### PR TITLE
Add Hermitian Quadrature

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Needs https://github.com/gonum/mathext/pull/1

Sorry @vladimir-ch , one more of these. This is why the Airy PR exists in the first place.  These tests are also a confirmation that the Airy code is doing the right thing.

A lot to clean up here as well to make more like go and less like matlab. This will happen once Airy is merged and the Fortran removed.

@kortschak Did I do the copyright stuff reasonably well?

Thanks @atj60gaibb @chebfun for the code/algorithms!

**Comment by kortschak** [comment]
Copyright notices and imported license are fine.

File: hermite.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+import (
+	"math"
+	"math/cmplx"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/mathext/airy"
+)
+
+// Hermite integrates a function over the whole real line with negative
```
Comment (vladimir-ch) [inline hermite.go]: Stricly speaking, Hermite does not _integrate_, it generates or computes sample points and weights for the Hermite quadrature. Perhaps the comment could be adjusted to reflect this.

Also, wouldn't GaussHermite be a more accurate, albeit longer, name?

File: hermite.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+// squared exponential weight
+//  int_-inf^inf e^-x^2 f(x) dx .
```
Comment (vladimir-ch) [inline hermite.go]: s/e^-x^2/exp(-x^2)/ ?

File: hermite.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+	if !math.IsInf(min, -1) || !math.IsInf(max, 1) {
+		panic("hermite: non-infinite bound")
+	}
+	h.locations(x, weight)
+}
+
+func (h Hermite) locations(x, weights []float64) {
+	n := len(x)
+	switch {
+	default:
```
Comment (vladimir-ch) [inline hermite.go]: This is strange: with the other two cases there is nothing left for the default case, the slice length will not be negative, and if the length is 0, the index will become -1. Change to:
```
	switch {
	case 0 < n && n <= 200:
		copy(x, xCacheHermite[n-1])
		copy(weights, wCacheHermite[n-1])
	case n > 200:
		xasy, weightsasy := h.locationsAsy(n)
		copy(x, xasy)
		copy(weights, weightsasy)
	}
```
? Also depends on what we want to do when n == 0.

File: hermite.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+		panic("n<0")
+	case n <= 200:
+		copy(x, xCacheHermite[n-1])
+		copy(weights, wCacheHermite[n-1])
+	case n > 200:
+		xasy, weightsasy := h.locationsAsy(n)
+		copy(x, xasy)
+		copy(weights, weightsasy)
+	}
+	return
```
Comment (vladimir-ch) [inline hermite.go]: Delete

File: hermite.go [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+	// need to store all of that data.
+
+	// References:
+	// Algorithm:
+	// G. H. Golub and J. A. Welsch, "Calculation of Gauss quadrature rules",
+	// Math. Comp. 23:221-230, 1969.
+	// A. Glaser, X. Liu and V. Rokhlin, "A fast algorithm for the
+	// calculation of the roots of special functions", SIAM Journal
+	// on Scientific Computing", 29(4):1420-1438:, 2007.
+	// A. Townsend, T. Trogdon and S. Olver, Fast computation of Gauss
```
Comment (vladimir-ch) [inline hermite.go]:
```
A. Townsend, T. Trogdon, and S.Olver, Fast computation of Gauss quadrature nodes and weights
on the whole real line, IMA J. Numer. Anal., 36: 337–358, 2016. http://arxiv.org/abs/1410.5286
```

File: hermite.go [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+func (h Hermite) locationsAsy(n int) (x, w []float64) {
+	// A. Townsend, T. Trogdon and S. Olver, Fast computation of Gauss
```
Comment (vladimir-ch) [inline hermite.go]:
```
A. Townsend, T. Trogdon, and S.Olver, Fast computation of Gauss quadrature nodes and weights
on the whole real line, IMA J. Numer. Anal., 36: 337–358, 2016. http://arxiv.org/abs/1410.5286
```

File: hermite.go [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+		t0 := math.Cos(t)
+		xi := math.Sqrt(2*float64(n)+1) * t0
+		x[i] = xi
+		ders := xi*val[i] + math.Sqrt2*dval[i]
+		w[i] = math.Exp(-xi*xi) / (ders * ders)
+	}
+	return x, w
+}
+
+// hermpolyAsyAiry evaluates the hermite polynomials using the Airy asymptotic
```
Comment (vladimir-ch) [inline hermite.go]: Hermite

File: hermite.go [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+// formula in theta-space
```
Comment (vladimir-ch) [inline hermite.go]: Dot.

File: hermite.go [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+	// Initial guesses for Hermite zeros.
+	//
+	// [1] L. Gatteschi, Asymptotics and bounds for the zeros of Laguerre
+	// polynomials: a survey, J. Comput. Appl. Math., 144 (2002), pp. 7-27.
+	//
+	// [2] F. G. Tricomi, Sugli zeri delle funzioni di cui si conosce una
+	// rappresentazione asintotica, Ann. Mat. Pura Appl. 26 (1947), pp. 283-300.
+
+	// Gatteschi formula involving airy roots [1].
+	// These initial guess are good near x = sqrt(n+1/2);
```
Comment (vladimir-ch) [inline hermite.go]: guesses and dot at the end.

**Comment by vladimir-ch** [comment]
Nice. LGTM after comments addressed.

File: hermite.go (hunk 1 above)
Comment (btracey) [inline hermite.go]: Is there another Hermite in this context? It should also be GaussLegendre, GaussJacobi, etc. The Gauss seems redundant

File: hermite.go (hunk 2 above)
Comment (btracey) [inline hermite.go]: Done.

File: hermite.go (hunk 1 above)
Comment (vladimir-ch) [inline hermite.go]: I don't think so. Hermite is ok.

File: hermite.go (hunk 3 above)
Comment (btracey) [inline hermite.go]: Thanks, done.

File: hermite.go (hunk 4 above)
Comment (btracey) [inline hermite.go]: Done.

File: hermite.go (hunk 6 above)
Comment (btracey) [inline hermite.go]: Done.

File: hermite.go (hunk 7 above)
Comment (btracey) [inline hermite.go]: Done.

File: hermite.go (hunk 8 above)
Comment (btracey) [inline hermite.go]: Done.

File: hermite.go (hunk 9 above)
Comment (btracey) [inline hermite.go]: Done.

**Pull request merged**